proof/refine/ARM/Schedule_R.thy

```diff
@@ -2385,20 +2385,23 @@ global_interpretation refillPopHead: typ_at_all_props' "refillPopHead scPtr"
 context begin interpretation Arch . (*FIXME: arch_split*)
 
 lemma mergeRefills_valid_objs':
-  "\<lbrace>\<lambda>s. valid_objs' s \<and> active_sc_at' scPtr s \<and> obj_at' (\<lambda>sc'. Suc 0 < scRefillCount sc') scPtr s\<rbrace>
+  "\<lbrace>\<lambda>s. valid_objs' s \<and> sc_at' scPtr s \<and> ((\<lambda>sc. 1 < scRefillCount sc) |< scs_of' s) scPtr\<rbrace>
    mergeRefills scPtr
    \<lbrace>\<lambda>_. valid_objs'\<rbrace>"
   apply (clarsimp simp: mergeRefills_def updateRefillHd_def)
-  apply (rule_tac B="\<lambda>_ s. valid_objs' s \<and> active_sc_at' scPtr s" in hoare_seq_ext[rotated])
+  apply (rule_tac B="\<lambda>_ s. valid_objs' s \<and> sc_at' scPtr s" in hoare_seq_ext[rotated])
    apply wpsimp
+   apply (clarsimp simp: obj_at_simps opt_map_red)
   apply (wpsimp wp: set_sc_valid_objs' simp: updateSchedContext_def)
-  apply (clarsimp simp: is_active_sc'_def valid_sched_context'_def obj_at_simps valid_obj'_def
-                        valid_sched_context_size'_def opt_map_red active_sc_at'_rewrite
+  apply (clarsimp simp: valid_sched_context'_def obj_at_simps valid_obj'_def
+                        valid_sched_context_size'_def opt_map_red
                  dest!: sc_ko_at_valid_objs_valid_sc')
   done
 
-crunches mergeRefills, nonOverlappingMergeRefills
+crunches mergeRefills, nonOverlappingMergeRefills, scheduleUsed
   for active_sc_at'[wp]: "active_sc_at' scPtr"
+  and sc_at'[wp]: "\<lambda>s. P (sc_at' scPtr s)"
+  (simp: objBits_simps)
```

Comment:
Tiny nitpick: this one and a later `sc_at'` crunch should probably be in terms of `typ_at'` and `typ_at_all_props'`

Reply:
Doesn’t `crunch` generate more lemmas about `sc_at’` in this case? The proofs below use not just the lemmas for the top level functions (i.e., mergeRefills, nonOverlappingMergeRefills, scheduleUsed) but also those for “sub definitions”, so it is a bit of pain to list all locale interpretations needed. They do generate all other good stuff, though. Also, it would be nice if we can do 2-dimensional locale interpretation like `crunches`.

Reply:
I added a bunch of interpretations and tweaked the crunches. Crunches are better done on `refillUnblockCheck` and `refillBudgetCheck` and put together before the interpretations.

Reply:
Looks good to me! https://sel4.atlassian.net/browse/VER-1371 would definitely be nice to have in this case with all of those interpretations.